packages/contracts-bedrock/contracts/test/invariants/OptimismPortal.t.sol

```diff
@@ -0,0 +1,165 @@
+pragma solidity 0.8.15;
+
+import { InvariantTest } from "forge-std/InvariantTest.sol";
+import { Portal_Initializer } from "../CommonTest.t.sol";
+import { Types } from "../../libraries/Types.sol";
+
+contract OptimismPortal_Invariant_Harness is Portal_Initializer, InvariantTest {
+    // Reusable default values for a test withdrawal
+    Types.WithdrawalTransaction _defaultTx;
+
+    uint256 _proposedOutputIndex;
+    uint256 _proposedBlockNumber;
+    bytes32 _stateRoot;
+    bytes32 _storageRoot;
+    bytes32 _outputRoot;
+    bytes32 _withdrawalHash;
+    bytes[] _withdrawalProof;
+    Types.OutputRootProof internal _outputRootProof;
+
+    function setUp() public virtual override {
```

Comment:
Just to make sure I understand

For each round:
- `setUp` is called
- A number of calls are made to the `targetContract`
- Each invariant is checked

Reply:
Yep! A high-level overview of how forge's invariant tests work behind the scenes can be found [here](https://github.com/mds1/solidity-sandbox/blob/main/test/4_InvariantNonceGoUp.t.sol#L19-L30)